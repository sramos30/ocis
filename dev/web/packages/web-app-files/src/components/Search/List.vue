<template>
  <div class="files-search-result">
    <app-bar :has-bulk-actions="true" />
    <no-content-message v-if="!paginatedResources.length" class="files-empty" icon="folder">
      <template #message>
        <p class="oc-text-muted">
          <span v-if="!!$route.query.term" v-translate>No resource found</span>
          <span v-else v-translate>No search term entered</span>
        </p>
      </template>
    </no-content-message>
    <resource-table
      v-else
      v-model="selectedResources"
      class="files-table"
      :class="{ 'files-table-squashed': false }"
      :resources="paginatedResources"
      :target-route="resourceTargetLocation"
      :are-paths-displayed="true"
      :are-thumbnails-displayed="displayThumbnails"
      :has-actions="true"
      :is-selectable="false"
      @fileClick="$_fileActions_triggerDefaultAction"
      @rowMounted="rowMounted"
    >
      <template #contextMenu="{ resource }">
        <context-actions v-if="isResourceInSelection(resource)" :items="selectedResources" />
      </template>
      <template #footer>
        <pagination :pages="paginationPages" :current-page="paginationPage" />
        <list-info
          v-if="paginatedResources.length > 0"
          class="oc-width-1-1 oc-my-s"
          :files="totalFilesCount.files"
          :folders="totalFilesCount.folders"
          :size="totalFilesSize"
        />
      </template>
    </resource-table>
  </div>
</template>

<script lang="ts">
import { useResourcesViewDefaults } from '../../composables'
import { VisibilityObserver } from 'web-pkg/src/observer'
import { ImageType, ImageDimension } from '../../constants'
import { createLocationSpaces } from '../../router'
import NoContentMessage from 'web-pkg/src/components/NoContentMessage.vue'
import ResourceTable from '../FilesList/ResourceTable.vue'
import ContextActions from '../FilesList/ContextActions.vue'
import debounce from 'lodash-es/debounce'
import { mapMutations, mapGetters, mapActions } from 'vuex'
import AppBar from '../AppBar/AppBar.vue'
import { defineComponent } from '@vue/composition-api'
import ListInfo from '../FilesList/ListInfo.vue'
import Pagination from '../FilesList/Pagination.vue'
import MixinFileActions from '../../mixins/fileActions'
import MixinFilesListFilter from '../../mixins/filesListFilter'
import MixinFilesListScrolling from '../../mixins/filesListScrolling'
import { Resource } from '../../helpers/resource'

const visibilityObserver = new VisibilityObserver()

export default defineComponent({
  components: { AppBar, ContextActions, ListInfo, Pagination, NoContentMessage, ResourceTable },
  mixins: [MixinFileActions, MixinFilesListFilter, MixinFilesListScrolling],
  props: {
    searchResults: {
      type: Array,
      default: function () {
        return []
      }
    }
  },
  setup() {
    return {
      ...useResourcesViewDefaults<Resource, any, any[]>(),

      resourceTargetLocation: createLocationSpaces('files-spaces-personal-home')
    }
  },
  computed: {
    ...mapGetters(['configuration']),
    ...mapGetters('Files', ['totalFilesCount', 'totalFilesSize']),
    displayThumbnails() {
      return !this.configuration.options.disablePreviews
    }
  },
  watch: {
    searchResults: {
      handler: function () {
        this.CLEAR_CURRENT_FILES_LIST()
        this.LOAD_FILES({
          currentFolder: null,
          files: this.searchResults.length
            ? this.searchResults.map((searchResult) => searchResult.data)
            : []
        })
      },
      immediate: true
    }
  },
  beforeDestroy() {
    visibilityObserver.disconnect()
  },
  methods: {
    ...mapMutations('Files', ['CLEAR_CURRENT_FILES_LIST', 'LOAD_FILES']),
    ...mapActions('Files', ['loadPreview']),
    rowMounted(resource, component) {
      if (!this.displayThumbnails) {
        return
      }

      const debounced = debounce(({ unobserve }) => {
        unobserve()
        this.loadPreview({
          resource,
          isPublic: false,
          dimensions: ImageDimension.Thumbnail,
          type: ImageType.Thumbnail
        })
      }, 250)

      visibilityObserver.observe(component.$el, { onEnter: debounced, onExit: debounced.cancel })
    }
  }
})
</script>
