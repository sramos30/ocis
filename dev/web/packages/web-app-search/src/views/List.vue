<template>
  <div class="oc-width-1-1">
    <div
      v-if="$asyncComputed.searchResults.updating"
      class="oc-flex oc-flex-middle oc-flex-center oc-height-1-1 oc-width-1-1"
    >
      <oc-spinner size="large" :aria-hidden="true" aria-label="" />
    </div>
    <template v-else>
      <component :is="listSearch.component" :search-results="searchResults" />
    </template>
  </div>
</template>

<script>
import { providerStore } from '../service'

export default {
  data() {
    const { provider: providerId } = this.$route.query
    const { listSearch } = providerStore.availableProviders.find(
      (provider) => provider.id === providerId
    )
    // abort and return if no provider is found
    return {
      listSearch
    }
  },

  asyncComputed: {
    searchResults: {
      get() {
        return this.listSearch.search(this.$route.query.term)
      },
      watch: ['$route.query.term', '$route.query.provider']
    }
  }
}
</script>

<style lang="scss">
#search-view {
  padding: 25px 10px 10px 10px;
}
</style>
