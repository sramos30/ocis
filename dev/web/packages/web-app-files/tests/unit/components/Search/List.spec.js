import { mount, createLocalVue } from '@vue/test-utils'
import VueCompositionAPI from '@vue/composition-api'
import VueRouter from 'vue-router'
import Vuex from 'vuex'
import DesignSystem from 'owncloud-design-system'
import GetTextPlugin from 'vue-gettext'

import List from '@files/src/components/Search/List.vue'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VueRouter)
localVue.use(DesignSystem)
localVue.use(VueCompositionAPI)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const stubs = {
  'app-bar': true,
  'no-content-message': false,
  'resource-table': false,
  pagination: true,
  'list-info': true
}

const selectors = {
  noContentMessage: '.files-empty',
  filesTable: '.files-table',
  pagination: 'pagination-stub',
  listInfo: 'list-info-stub'
}

describe('List component', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  describe.each(['no search term is entered', 'no resource is found'])('when %s', (message) => {
    let wrapper
    beforeEach(() => {
      if (message === 'no search term is entered') {
        wrapper = getWrapper()
      } else {
        wrapper = getWrapper('epsum.txt')
      }
    })

    it('should show no-content-message component', () => {
      const noContentMessage = wrapper.find(selectors.noContentMessage)

      expect(noContentMessage.exists()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })
    it('should not show files table', () => {
      const filesTable = wrapper.find(selectors.filesTable)
      const listInfo = wrapper.find(selectors.listInfo)

      expect(filesTable.exists()).toBeFalsy()
      expect(listInfo.exists()).toBeFalsy()
    })
  })

  describe('when resources are found', () => {
    const spyTriggerDefaultAction = jest
      .spyOn(List.mixins[0].methods, '$_fileActions_triggerDefaultAction')
      .mockImplementation()
    const spyRowMounted = jest.spyOn(List.methods, 'rowMounted')

    let wrapper
    beforeEach(() => {
      wrapper = getWrapper('lorem', files)
    })

    it('should not show no-content-message component', () => {
      const noContentMessage = wrapper.find(selectors.noContentMessage)

      expect(noContentMessage.exists()).toBeFalsy()
    })
    it('should set correct props on list-info component', () => {
      const listInfo = wrapper.find(selectors.listInfo)

      expect(listInfo.exists()).toBeTruthy()
      expect(listInfo.props().files).toEqual(files.length)
      expect(listInfo.props().folders).toEqual(0)
      expect(listInfo.props().size).toEqual(getTotalSize(files))
    })
    it('should trigger the default action when a "fileClick" event gets emitted', async () => {
      const filesTable = wrapper.find(selectors.filesTable)

      expect(spyTriggerDefaultAction).toHaveBeenCalledTimes(0)

      await filesTable.vm.$emit('fileClick')

      expect(spyTriggerDefaultAction).toHaveBeenCalledTimes(1)
    })
    it('should lazily load previews when a "rowMounted" event gets emitted', () => {
      expect(spyRowMounted).toHaveBeenCalledTimes(files.length)
    })
  })
})

function getWrapper(searchTerm = '', files = []) {
  return mount(List, {
    localVue,
    propsData: {
      searchResults: getSearchResults(files)
    },
    store: createStore(files),
    router: new VueRouter(),
    stubs
  })
}

function createStore(activeFiles) {
  return new Vuex.Store({
    getters: {
      configuration: () => ({
        options: {
          disablePreviews: true
        }
      })
    },
    modules: {
      Files: {
        namespaced: true,
        getters: {
          activeFiles: () => activeFiles,
          selectedFiles: () => [],
          totalFilesCount: () => ({ files: activeFiles.length, folders: 0 }),
          totalFilesSize: () => getTotalSize(activeFiles),
          inProgress: () => [],
          currentFolder: () => {
            return {
              path: '',
              canCreate() {
                return false
              }
            }
          }
        },
        mutations: {
          CLEAR_CURRENT_FILES_LIST: jest.fn(),
          CLEAR_FILES_SEARCHED: jest.fn(),
          LOAD_FILES: jest.fn()
        }
      }
    }
  })
}

function getSearchResults(files) {
  return files.map((file) => ({ data: file, id: file.id }))
}

function getTotalSize(files) {
  return files.reduce((total, file) => total + file.size, 0)
}

const files = [
  {
    id: '1',
    path: 'lorem.txt',
    size: 100
  },
  {
    id: '2',
    path: 'lorem.pdf',
    size: 50
  }
].map((file) => {
  return {
    ...file,
    canDownload: () => true,
    canBeDeleted: () => true,
    isReceivedShare: () => false,
    isMounted: () => false
  }
})
