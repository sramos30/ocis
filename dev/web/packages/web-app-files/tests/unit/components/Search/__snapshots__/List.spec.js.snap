// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component when no resource is found should show no-content-message component 1`] = `
<div class="files-search-result">
  <app-bar-stub breadcrumbs="" breadcrumbscontextactionsitems="" hasbulkactions="true" hassidebartoggle="true" hasviewoptions="true"></app-bar-stub>
  <div class="oc-height-1-1 oc-flex oc-flex-column oc-flex-center oc-flex-middle oc-text-center files-empty">
    <div class="oc-icon oc-icon-xxl oc-icon-passive">
      <!---->
    </div>
    <div class="oc-text-muted oc-text-large">
      <p class="oc-text-muted"><span data-msgid="No search term entered" data-current-language="en_US">No search term entered</span></p>
    </div>
    <div class="oc-text-muted"></div>
  </div>
</div>
`;

exports[`List component when no search term is entered should show no-content-message component 1`] = `
<div class="files-search-result">
  <app-bar-stub breadcrumbs="" breadcrumbscontextactionsitems="" hasbulkactions="true" hassidebartoggle="true" hasviewoptions="true"></app-bar-stub>
  <div class="oc-height-1-1 oc-flex oc-flex-column oc-flex-center oc-flex-middle oc-text-center files-empty">
    <div class="oc-icon oc-icon-xxl oc-icon-passive">
      <!---->
    </div>
    <div class="oc-text-muted oc-text-large">
      <p class="oc-text-muted"><span data-msgid="No search term entered" data-current-language="en_US">No search term entered</span></p>
    </div>
    <div class="oc-text-muted"></div>
  </div>
</div>
`;
